### DISCUSSION TITLE
Vue support

### DISCUSSION THREAD

**Discussion by loganvolkers** [pmndrs/jotai#1811] [Ideas]
Jotai seems better than [nanostores](https://github.com/nanostores/nanostores) from our experience using it, but it is [recommended by Astro](https://docs.astro.build/en/core-concepts/sharing-state/) for sharing state across frameworks. I think this is due to nanostores cross-framework support.

We'd like to start using Jotai with other frameworks, and I think we can look at [nanostores vue](https://github.com/nanostores/vue) as a baseline for building a vue integration. It feels like out-of-the-box adapters should be a priority for boosting jotai adoption.

**Comment by dai-shi** [comment]
Sounds interesting! Would you be working on it?
We can host it under https://github.com/jotaijs/ if you prefer.

fyi, https://github.com/wobsoriano/solid-jotai is developed by @wobsoriano.

**Comment by wobsoriano** [comment]
I forgot about the Vue integration but I started it before - https://github.com/wobsoriano/jotai-vue

I focused on the solid integration

**Comment by loganvolkers** [comment]
Related to this would be a [Pinia](https://pinia.vuejs.org/) integration.

Looks like it's possible using `store.$subscribe`: https://pinia.vuejs.org/core-concepts/state.html#subscribing-to-the-state

**Reply by wobsoriano** [reply]
Curious as to how would Pinia integrate with Jotai?

**Reply by loganvolkers** [reply]
Similar to the integrations with [Redux](https://jotai.org/docs/integrations/redux), [Xstate](https://jotai.org/docs/integrations/xstate), [Zustand](https://jotai.org/docs/integrations/zustand), [TanStack Query](https://jotai.org/docs/integrations/query), [urql](https://github.com/urql-graphql/urql) or any other state management solution.

This would be useful if you're trying to mix state management solutions (e.g. combine Pinia and Redux, or Pinia and TanStack Query) or are working across frameworks. 

(Aside: In my opinion TanStack Query should be built on atoms, instead of its custom state management primitives.)

You would define an `atomFromPina` that takes in a store, listens to changes, and delegates actions.

```js
// WORK IN PROGRESS - example
function atomFromPina(store){
  // TODO: Probably needs `mapActions`
  const baseAtom = atom(store.state);

  baseAtom.onMount = (setAtom)=>{
      return store.$subscribe((mutation,state)=>setAtom(state),  { detached: true } );
  }
  
  return atom(
     // Listen to changes
     get=>get(baseAtom),
     // Delegates updates to Pinia store
     // TODO: Probably needs `mapActions`
     (get,set,cb)=>cb(store)
   );
}
```

And then you could use a store, for example:
```js
import { defineStore } from 'pinia'

const useCounterStore = defineStore('counter', {
  state: () => {
    return { count: 0 }
  },
  // could also be defined as
  // state: () => ({ count: 0 })
  actions: {
    increment() {
      this.count++
    },
  },
})

const pinaAtom = atomFromPina(useCounterStore);
const doublePiniaAtom = atom(get=>get(pinaAtom).count*2);
const incrementPiniaAtom = atom(null,()=>set(pinaAtom, x=>x.increment() ));
```

**Comment by AjaxSolutions** [comment]
> Jotai seems better than [nanostores](https://github.com/nanostores/nanostores) from our experience

@loganvolkers Could you please briefly describe why Jotai is better than Nano Stores from your experience?

**Reply by loganvolkers** [reply]
> Could you please briefly describe why Jotai is better than Nano Stores from your experience?

- Ecosystem: Jotai has integrations with other state libraries. It's the polyglot store framework that lets you connect redux, zustand, and valtio across a single app. 
- Maintainability: Jotai lets you replace a primitive atom with a derived atom, and no downstream atoms need to be updated. This allows for easier refactoring, particularly for library authors. For example, it is easy to replace `atom` with `atomWithStorage` without other app changes.
- Adoption: Jotai has more downloads. That momentum means something.

On the flip side, Nanostores wins for their simple imperative API and multi framework support.

**Comment by ortonomy** [comment]
I'd love to have vue support for Jotai, simply because I'm so familiar with Jotai now, it's a fluent and easy to grok API, the support for things like Immer, Tanstack Query etc is great... so being able to drag it into my Vue projects would be a boon as it's now my go to - no more need for context, redux or anything else so ... bulky.

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/discussions/1811#discussioncomment-5197808
